features/generate.feature

```diff
@@ -111,3 +111,9 @@ Scenario: The project is an iOS application that has resources (ios_app_with_cus
     Then I should be able to test the scheme Framework1Tests
     Then I should be able to build the scheme Framework2
     Then I should be able to test the scheme Framework2Tests
+
+Scenario: The project is an iOS application that has multiple dependencies (ios_app_with_multiple_dependencies)
```

Comment:
I'd remove this test entirely. We are already testing the sorting logic with unit tests. The last step here is testing that the objects always get the same hash, which is something that is already tested by XcodeProj.

Reply:
I couldn't find any tests verifying it, eh, just realised that my description is quite misleading in XcodeProj context.

The test case gets the checksum of the whole project.pbxproj file, so the test ensures the whole generated project file is bit-by-bit identical with the one generated before. This test would catch both issues (embeddedFramework and products).

Reply:
I see what you mean. In that case, it makes sense to have a test but a unit test one. The `ProjectFileElements` is responsible for, given a project, create a project with files and groups structure, right? You can give a project as an input and then verify that the output Xcode project has always the same files listed in the same order.

It's hard to draw a line between what should be a unit test and what should be an acceptance test. I'd say whenever it's possible, write a unit test *(like in this case)*, and leave acceptance tests for user use cases like: Hey, I want to start using Tuist so I generate my project and I expect it to compile.

Does that make sense?

Reply:
@pepibumur does the rewording of the test case as I suggested help clarify this case? The use case we wanted to validate was re-generating the the same project several times always yielded an identical Xcode project file (not so much the details of the hashes internally). This would help eliminate the git diff that appears when regenerating projects that are checked into source control that haven't actually changed.

I agree it's an interesting one as to where to draw the line between acceptance test / unit test. Perhaps an alternate is an integration test in `TuistIntegrationTests` as final Xcode project file is a result of several components working together. 

(Now I think of it, will an integration test catch this? If I'm not mistaken, Swift utilises a random seed per process for the hash values - and the Integration tests run as a single Swift process)

Reply:
> @pepibumur does the rewording of the test case as I suggested help clarify this case? 

A bit. I think it makes sense to test whether the structure of the project is the same every time we call generate. We can easily verify that by taking advantage of the `Equatable` conformance of the [`XcodeProj`](https://github.com/tuist/xcodeproj/blob/master/Sources/xcodeproj/Project/XcodeProj.swift#L5) model. That can be a test in the `TuistIntegrationTests` target.

The current acceptance test not only tests that but implicitly tests that XcodeProj *(the library)* is doing the right job generating those hashes. Some developers might think that's a good idea, but I'm more in the group that thinks that you should not test whether other libraries or components are doing their job. XcodeProj has already [tests](https://github.com/tuist/xcodeproj/blob/master/Tests/xcodeprojTests/Objects/Project/PBXProjIntegrationTests.swift#L35) that verify that.